runtime/lua/vim/_helptags.lua

```diff
@@ -0,0 +1,164 @@
+local M = {}
+
+local ts = vim.treesitter
+local query = ts.query.parse('vimdoc', '(tag (word) @tagname)')
+
+--- @alias TagLocation { [1]: string, [2]: string, [3]: string}[] tuple of tag, file, and search command
+
+--- Find all helptags in a single file.
+---
+--- @param filename string helpfile with tags
+--- @return TagLocation[] # empty if file is not readable
+local function extract_file_tags(filename)
+  local f, err = io.open(filename, 'r')
+  if f == nil or err then
+    vim.notify(('E153: Unable to open %s for reading'):format(filename), vim.log.levels.ERROR)
+    return {}
+  end
+  local source = f:read('*a')
+  f:close()
+  local fn = vim.fs.basename(filename)
+
+  local tags = {}
+  local tree = ts.get_string_parser(source, 'vimdoc'):parse()[1]:root()
+
+  for _, match in query:iter_matches(tree, source) do
+    for id, node in pairs(match) do
+      if query.captures[id] == 'tagname' then
+        local tagname = ts.get_node_text(node[1], source)
+        local escaped = tagname:gsub('[\\/]', '\\%0')
+        local searchcmd = '/*' .. escaped .. '*'
+        table.insert(tags, { tagname, fn, searchcmd })
+      end
+    end
+  end
+
+  return tags
+end
+
+--- Report duplicate tags.
+---
+--- @param tags TagLocation[]
+--- @return boolean # true if there are duplicate tags
+local function duplicate_tags(tags)
+  local found = false
+  local prevtag, prevfn
+  for _, tag in ipairs(tags) do
+    local curtag, curfn, _ = unpack(tag)
+    if curtag == prevtag then
+      found = true
+      local other_fn = prevfn ~= curfn and (' and ' .. prevfn) or ''
+      vim.notify(
+        ('E154: Duplicate tag "%s" in %s%s'):format(curtag, curfn, other_fn),
+        vim.log.levels.WARN
+      )
+    end
+    prevtag = curtag
+    prevfn = curfn
+  end
+  return found
+end
+
+--- Extract tags from a list of helpfiles.
+---
+--- @param helpfiles string[] list of helpfiles
+--- @param tagsfile string the filename of the 'tags' file
+--- @param include_helptags_tag boolean true if the 'tags' tag should be included
+local function create_tags_from_files(helpfiles, tagsfile, include_helptags_tag)
+  ---@type TagLocation
+  local tags = {}
+  local i = 1
+  local function process_helpfile(co)
+    if i > #helpfiles then
+      coroutine.resume(co, true)
+      return
+    end
+
+    local filename = helpfiles[i]
+    i = i + 1
+
+    vim.schedule(function()
+      local filetags = extract_file_tags(filename)
+      vim.list_extend(tags, filetags)
+      process_helpfile(co)
+    end)
+  end
+
+  coroutine.wrap(function()
```

Comment:
I don't think we should add manual async at this point since we're already considering a `vim.async` module based on @lewis6991 work. Please look at #34009 and https://github.com/lewis6991/async.nvim, and pull in whatever you need from that into a `vim._async` module (that we'll merge with the one from #34009).

Reply:
The potential of `vim.async` shouldn't restrict the use of coroutines. `vim.async` is just a very generalized layer on top of coroutines, but if the requirements are small (like) here, then I don't see a problem with using coroutines. I still think coroutines are important for generator functions (functions that return iterators).

The usage here is contained within ~30LOC, if anything this is might be preferable to using `vim.async`.

Though I'm not quite sure why coroutines are being used here (needs a comment). It looks like it's to not eat up the main loop, in which case `vim.async` + `uv.fs_open` might help in the future.

What is the effect if no coroutines where used here? Does creating tags from helpfiles cause lots of lag or something?

Reply:
I noticed it blocked the editor (~2 secs for the runtimedir). Although maybe vim.schedule will be enough for this. 

I added coroutines because I was looking for parallelization, as extracting tags for every file is a task that can be parallelized very efficiently. But I learned that coroutines all run on a single thread and file reads are blocking. Still, this implementation felt 'snappier' and faster to me than the fully sync approach. I'll try vim.schedule (+ uv.fs_open)